File: PathSum/II/PathSumII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        res = []
        if root is None:
            return res
        Sum(root, res, [], sum, 0)
        return res

def Sum(root, res, tmp, sum, t_sum):
    if root is None:
        return
    t_sum = t_sum + root.val
    if root.left is None and root.right is None:
        if t_sum == sum:
            tmp.append(root.val)
            res.append(tmp)
            return
        else:
            return
    Sum(root.left, res, tmp + [root.val], sum, t_sum)
    Sum(root.right, res, tmp + [root.val], sum, t_sum)

File: PathSum/II/PathSumII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> result;
        if(root == nullptr)
        	return result;
        vector<int> subSum;
        BACK(result, subSum, sum, root);
        return result;
    }
private:
	void BACK(vector<vector<int>>& result, vector<int>& subSum, int sum, TreeNode* node) {
		if(node->left == nullptr && node->right == nullptr) {
			if(sum == node->val) {
				subSum.push_back(node->val);
				result.push_back(subSum);
				subSum.pop_back();
			}
			return;
		}
		subSum.push_back(node->val);
		if(node->left)
			BACK(result, subSum, sum - node->val, node->left);
		if(node->right)
			BACK(result, subSum, sum - node->val, node->right);
		subSum.pop_back();
		return;
	}
};